File: PAT_Basic_Level_by_cpp/1054 求平均值/1054 求平均值.cpp
// 1054 求平均值
//
// 实现原理：
// 从 N 个字符串中筛选出合法的实数，计算它们的平均值。合法数值需满足：
//   1. 在 [-1000, 1000] 范围内
//   2. 最多包含两位小数
// 非法字符串逐条输出错误信息，最后根据合法个数输出平均值。
//
// 校验函数 isValid 的算法：
//   逐字符扫描字符串，按状态机方式解析：
//   - 允许以 '+' 或 '-' 开头（但不能只有符号）
//   - 允许一个小数点，小数点后最多两位数字
//   - 其余字符必须为数字
//   若通过格式校验则调用 stod 转为 double，再检查值域范围。
//
// 复杂度分析：
// 时间复杂度：O(N*L)，其中 N 为字符串个数，L 为字符串平均长度。
// 空间复杂度：O(1)，只使用常数额外空间。
#include <iostream>
#include <string>
#include <cctype>
#include <cstdio>

using namespace std;

// 校验字符串是否为合法数值，并将值写入 val
// 合法条件：[-1000,1000] 范围内，最多两位小数
bool isValid(const string& s, double& val) {
    if (s.empty()) return false;
    size_t i = 0;
    // 处理可选的符号位
    if (s[0] == '+' || s[0] == '-') {
        if (s.size() == 1) return false;  // 只有符号无数字
        i = 1;
    }
    int digits = 0, afterDot = 0;
    bool hasDot = false;
    for (; i < s.size(); ++i) {
        if (s[i] == '.') {
            if (hasDot) return false;  // 多个小数点，非法
            hasDot = true;
        } else if (isdigit((unsigned char)s[i])) {
            if (hasDot) ++afterDot;    // 小数点后的数字位数
            else ++digits;             // 小数点前的数字位数
        } else {
            return false;               // 出现非法字符
        }
    }
    if (afterDot > 2) return false; // 最多两位小数
    val = stod(s);                   // 字符串转浮点数
    if (val < -1000.0 || val > 1000.0) return false;  // 检查值域
    return true;
}

int main() {
    int n;
    if (!(cin >> n)) return 0;

    int cnt = 0;
    double sum = 0.0;
    for (int i = 0; i < n; ++i) {
        string s;
        cin >> s;
        double v;
        if (isValid(s, v)) {
            ++cnt;
            sum += v;
        } else {
            // 对非法输入逐条报错
            cout << "ERROR: " << s << " is not a legal number" << endl;
        }
    }

    // 根据合法个数输出不同格式的平均值
    if (cnt == 0)
        cout << "The average of 0 numbers is Undefined" << endl;
    else if (cnt == 1)
        printf("The average of 1 number is %.2f\n", sum);
    else
        printf("The average of %d numbers is %.2f\n", cnt, sum / cnt);

    return 0;
}

File: PAT_Basic_Level_by_lua/1054 求平均值/1054 求平均值.lua
-- 题目编号: 1054 求平均值
-- 实现原理: 逐个读取输入的字符串，用tonumber尝试转为数字，并用正则匹配校验是否为最多两位小数的合法格式。检查数值范围[-1000,1000]，合法则累加求和并计数；非法则输出错误信息。最后根据合法数字个数输出平均值，注意单复数及Undefined的处理。
-- n: 输入数字个数，sum: 合法数字之和，c: 合法数字个数
local n, sum, c = tonumber(io.read("*n")), 0, 0
local input = io.read("*a")
local values = {}
for s in input:gmatch("%S+") do
    values[#values + 1] = s
end
-- 逐个处理输入
for i = 1, n do
    local s = values[i]
    local x = tonumber(s)  -- 尝试转换为数字
    -- 校验：能转为数字 且 在[-1000, 1000]范围内 且 格式合法（最多两位小数）
    -- 正则说明：^[+-]? 可选正负号，%d+ 整数部分，%.?%d?%d? 可选小数部分（最多两位）
    local format_ok = s and (s:match("^[+-]?%d+$") or s:match("^[+-]?%d+%.%d%d?$"))
    if x and x >= -1000 and x <= 1000 and format_ok then
        sum, c = sum + x, c + 1  -- 累加求和并计数
    else
        -- 非法数字，输出错误信息
        print("ERROR: " .. s .. " is not a legal number")
    end
end
-- 输出结果
if c == 0 then
    -- 没有合法数字，输出Undefined
    print("The average of 0 numbers is Undefined")
else
    -- 输出平均值，注意单复数处理
    print(string.format("The average of %d number%s is %.2f", c, c == 1 and "" or "s", sum / c))
end
